### PR TITLE
refactor ps benchmark

### PR DESCRIPTION
**Pull request by gcramer23**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60785 add trainer hook functions
* **#60784 refactor ps benchmark**

This pr refactors the ps benchmark for modular trainers.

Differential Revision: [D29697291](https://our.internmc.facebook.com/intern/diff/D29697291)

### CODE REVIEW

File: benchmarks/distributed/rpc/parameter_server/launcher.py [hunk 1]
```diff
@@ -243,13 +360,31 @@ def run_benchmark(rank, model, data, args, config):
 
 
 def get_json_config(file_name, id):
-    with open(os.path.join(Path(__file__).parent, file_name), "r") as f:
```
Comment (gcramer23) [inline benchmarks/distributed/rpc/parameter_server/launcher.py:246]: I'll add this back when responding to comments. I didn't realize someone submitted a pr on the benchmark.

Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/launcher.py:246]: I fixed it.

File: benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py [hunk 2]
```diff
@@ -1,42 +1,27 @@
...
-    def __init__(self, rank, trainer_count, process_group, use_cuda_rpc, server_rref, backend, epochs):
+from .TrainerBase import TrainerBase
+
+
+class DdpTrainer(TrainerBase):
+
+    def __init__(
+        self,
+        rank,
+        trainer_count,
```
Comment (mrshenli) [inline benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py:11]: Having a long list of args is usually not user-friendly. Are all of the args here necessary? If the goal of this benchmark to test efficiency of PS vs non-PS, does it make sense to drop the customization knobs for optimizers and loss functions??

File: benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py [hunk 3]
```diff
@@ -1,42 +1,27 @@
...
+        process_group,
+        use_cuda_rpc,
+        server_rref,
+        backend, epochs,
+        preprocess_data,
+        create_criterion,
+        create_ddp_model,
+        lr,
+        create_optimizer,
+        HookState,
```
Comment (mrshenli) [inline benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py:21]: Why do we have a class name here? Do you mean sth like `hook_state_class`?

File: benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py [hunk 4]
```diff
@@ -1,42 +1,27 @@
...
-            r"""
-            A method that increments batch_number by 1.
-            """
-            self.batch_number += 1
-
-    def __init__(self, rank, trainer_count, process_group, use_cuda_rpc, server_rref, backend, epochs):
+from .TrainerBase import TrainerBase
+
+
+class DdpTrainer(TrainerBase):
```
Comment (mrshenli) [inline benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py:6]: Just wanna share an example of good API design for trainers: https://github.com/PyTorchLightning/pytorch-lightning

This is not 100% applicable here, but hopeful it could serve as an example.

File: benchmarks/distributed/rpc/parameter_server/data/DummyData.py [hunk 5]
```diff
@@ -37,7 +47,7 @@ def generate_input():
         self.input = generate_input()
         self.target = torch.randint(0, max_val, [input_samples])
         self.start = 0
-        self.end = max_val
+        self.end = input_samples
```
Comment (mrshenli) [inline benchmarks/distributed/rpc/parameter_server/data/DummyData.py:50]: Is this used anywhere?

**Review by mrshenli — COMMENTED** [review]
Three general comments 

1. let's just modularize components that we expect users to override (i.e., overriding those would have impact specific to PS training efficiency). 
2. let's try to improve the file structure a bit to avoid tiny and fragmented files. 
3. The rest of the PyTorch is usually very careful when introducing new arguments to APIs, because more arguments usually is more confusing. Let's try to apply the same spirit here as well.

File: benchmarks/distributed/rpc/parameter_server/criterion_functions/cel.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import torch.nn as nn
+
+
+def cel(rank, model):
```
[the remark this answers is not in the document]
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/criterion_functions/cel.py:4]: okay I'll hard code it.

File: benchmarks/distributed/rpc/parameter_server/data/DummyData.py [hunk 7]
```diff
@@ -14,6 +15,15 @@ def __init__(
         input_dim: int,
         sparsity_percentage: int
     ):
+        r"""
+        A data class that generates random data.
+        Args:
+            max_val (int): the maximum value for an element
+            input_samples (int): count of training samples
+            input_dim (int): number of elements in a sample
```
[the remark this answers is not in the document]
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/data/DummyData.py:23]: I see. I will update it to sample_length.

File: benchmarks/distributed/rpc/parameter_server/optimizer_functions/sgd_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+
+def sgd_optimizer(parameters, lr):
+    r"""
+    A function that creates a SGD optimizer for training.
+    Args:
+        parameters (iterable): iterable of parameters to optimize
+        lr (float): learning rate
+    """
+    return torch.optim.SGD(parameters, lr)
```
[the remark this answers is not in the document]
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/optimizer_functions/sgd_optimizer.py:11]: I was making everything modular. I will hardcode the optimizer.

File: benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py (hunk 2 above)
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py:11]: I dropped the customization for optimizers. The loss functions I think should stay because different models may require different loss functions. For example, I cannot use the fairseq BertCommonsense model with torch CrossEntropyLoss.

I'll drop learning rate.

File: benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py (hunk 3 above)
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py:21]: Yes I will fix this.

File: benchmarks/distributed/rpc/parameter_server/data/DummyData.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/data/DummyData.py:23]: Also, I will update to input_samples to sample_count.

File: benchmarks/distributed/rpc/parameter_server/data/DummyData.py (hunk 5 above)
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/data/DummyData.py]: I removed it.

File: benchmarks/distributed/rpc/parameter_server/preprocess_data_functions/preprocess_dummy_data.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
+def preprocess_dummy_data(rank, data):
```
[the remark this answers is not in the document]
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/preprocess_data_functions/preprocess_dummy_data.py]: Okay, I created a trainer directory. I merged the trainer class files into trainer.py. I also moved all files related to trainer configurations to the trainer directory.

[the remark this answers is not in the document]
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/preprocess_data_functions/preprocess_dummy_data.py]: I also merged parameter server class files into server.py.

File: benchmarks/distributed/rpc/parameter_server/trainer/criterions.py [hunk 10]
```diff
@@ -0,0 +1,11 @@
+import torch.nn as nn
+
+
+def cel(rank, model):
```
Comment (H-Huang) [inline benchmarks/distributed/rpc/parameter_server/trainer/criterions.py:4]: model isn't used

**Review by H-Huang — APPROVED** [review]

File: benchmarks/distributed/rpc/parameter_server/trainer/trainer.py [hunk 11]
```diff
@@ -175,3 +177,91 @@ def clear_metrics(self):
...
+        hook_state_class,
+        hook,
+        iteration_step
+    ):
+        r"""
+        A trainer that implements a DDP training algorithm using a simple hook that performs allreduce
+        using the process_group implementation.
+        Args:
+            rank (int): worker rank
+            trainer_count (int): count of trainer in the world
```
Comment (mrshenli) [inline benchmarks/distributed/rpc/parameter_server/trainer/trainer.py:205]: can rank and trainer_count be derived from `process_group`?

**Review by mrshenli — APPROVED** [review]

File: benchmarks/distributed/rpc/parameter_server/trainer/trainer.py (hunk 11 above)
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/trainer/trainer.py:205]: Yes.

File: benchmarks/distributed/rpc/parameter_server/trainer/criterions.py (hunk 10 above)
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/trainer/criterions.py:4]: I will remove it.

**Comment by gcramer23** [comment]
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29697291).
